### PR TITLE
Support parsing of key path subscripts on metatypes

### PR DESCRIPTION
**Pull request by ahoppen**
This is needed to support parsing eg.
https://github.com/swiftlang/swift/pull/73242/files#diff-a94f38549dc87ce27c438b255ae2f29c69cc1c2c5e5069f327c5c7fae7bc553dR667

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please test

File: Sources/SwiftParser/Types.swift [hunk 1]
```diff
@@ -260,7 +259,9 @@ extension Parser {
 
     var loopProgress = LoopProgressCondition()
     while self.hasProgressed(&loopProgress) {
-      if !stopAtFirstPeriod, self.at(.period) {
+      if allowMemberTypes || (self.at(.period) && self.peek(isAt: .keyword(.Type), .keyword(.Protocol))),
```
Comment (rintaro) [inline Sources/SwiftParser/Types.swift:262]: What about Nested types like: `\Foo.Bar.[0]`?

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Comment by ahoppen** [comment]
@swift-ci Please test Linux

**Comment by ahoppen** [comment]
@swift-ci Please test Linux

File: Sources/SwiftParser/Expressions.swift [hunk 2]
```diff
@@ -997,18 +997,20 @@ extension Parser {
...
     var loopProgress = LoopProgressCondition()
     while self.hasProgressed(&loopProgress) {
-      // Check for a [] or .[] suffix. The latter is only permitted when there
-      // are no components.
+      // Check for a [] or .[] suffix. The latter is only permitted after the type name. Since we don't know what
+      // constitutes a nested type reference and what constitutes a type's member, we can only disallow it in the parser
+      // after seeing the first non-property component.
       if self.at(TokenSpec(.leftSquare, allowAtStartOfLine: false))
-        || (components.isEmpty && self.at(.period) && self.peek(isAt: .leftSquare))
+        || (components.allSatisfy({ $0.component.is(RawKeyPathPropertyComponentSyntax.self) }) && self.at(.period)
```
Comment (rintaro) [inline Sources/SwiftParser/Expressions.swift:1012]: I found `\T.?.?` is currently accepted, it also need this check. 

Also, could this `components.allSatisfy()` check a flag outside the loop? Like

Comment (ahoppen) [reply Sources/SwiftParser/Expressions.swift]: Updated to disallow these key paths. I found a fun example along the way that’s valid 😭. This prevents us from diagnosing the `.` as an unexpected token, which would generally be nicer diagnostics.

```swift
struct Foo {
  var bar: Int?
}

infix operator .?.

func .?.(_ x: AnyKeyPath, _ a: Int) {}

var blah = 2
\Foo?.?.bar.?.blah
```

**Review by rintaro — APPROVED** [review]

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Comment by ahoppen** [comment]
@swift-ci Please test macOS

**Pull request merged**